sys/aarch64/evec.S

```diff
@@ -60,28 +60,19 @@
         mrs     x11, spsr_el1
         stp     x10, x11, [sp, #CTX_ELR]
 .if \el == 1
-        add     x18, sp, #CTX_SIZE
+        add     x10, sp, #CTX_SIZE
 .else
-        mrs     x18, sp_el0
+        mrs     x10, sp_el0
 .endif
-        stp      lr, x18, [sp, #CTX_LR]
+        stp      lr, x10, [sp, #CTX_LR]
         .cfi_rel_offset lr, CTX_LR
         .cfi_rel_offset sp, CTX_SP
-        load_pcpu x18
 .endm
 
+/* Load context from sp. */
 .macro  load_ctx el
         msr     daifset, #2
-.if \el == 0
-        load_pcpu x18
-        ldr     x18, [x18, #PCPU_CURTHREAD]
-        ldr     x18, [x18, #TD_UCTX]
-        mov     sp,  x18
-.endif
```

Comment:
Are you sure we don't need that code? I am not sure if userspace `sp` is the same as kernel `sp` in this situation (so we had to take it from some special place).

Reply:
See the comment in `mips/ebase.S`: `sp` points to the user exception frame if we came from `user_exc_enter` or `switch_to_userspace`, which are all the possible cases. The code I removed was needed because `kern_execve()` called `user_exc_leave()` directly, which would cause the stack pointer to point somewhere further down the stack. `switch_to_userspace()` adjusts the stack pointer so that it points to the user exception frame.

Reply:
Actually, there's another way to get to `user_exc_leave`: `thread_entry_setup()` sets up the kernel stack of a thread so that it returns to userspace immediately after being switched to. When that thread reaches `user_exc_leave`, its stack pointer will point at the user exception frame allocated at the bottom of the stack, so we're good.

Reply:
Thanks! Now I understand.